taichi/backends/codegen_opengl.cpp

```diff
@@ -0,0 +1,524 @@
+  {
+    bool linear_index = true;
+    for (int i = 0; i < (int)stmt->ptr.size(); i++) {
+      if (stmt->ptr[i].offset != i) {
+        linear_index = false;
+      }
+    }
+    if (stmt->same_source() && linear_index &&
+        stmt->width() == stmt->ptr[0].var->width()) {
+      auto ptr = stmt->ptr[0].var;
+      emit("const {} {}({});", opengl_data_type_name(stmt->element_type()),
+           stmt->raw_name(), ptr->raw_name());
+    } else {
+      TI_NOT_IMPLEMENTED;
+    }
+  }
+
+  void visit(LocalStoreStmt *stmt) override
+  {
+    emit("{} = {};", stmt->ptr->raw_name(), stmt->data->raw_name());
+  }
+
+  void visit(AllocaStmt *alloca) override
+  {
+    emit("{} {}(0);",
+        opengl_data_type_name(alloca->element_type()),
+        alloca->raw_name());
+  }
+
+  void visit(ConstStmt *const_stmt) override
+  {
+    TI_ASSERT(const_stmt->width() == 1);
+    emit("const {} {} = {};", opengl_data_type_name(const_stmt->element_type()),
+         const_stmt->raw_name(), const_stmt->val[0].stringify());
+  }
+
+  void visit(ArgLoadStmt *stmt) override
+  {
+    const auto dt = opengl_data_type_name(stmt->element_type());
+    if (stmt->is_ptr) {
+      emit("const {} {} = _arg_{}({}); // is_ptr", dt, stmt->raw_name(),
+          data_type_short_name(stmt->element_type()), stmt->arg_id);
+    } else {
+      emit("const {} {} = _arg_{}({});", dt, stmt->raw_name(),
+          data_type_short_name(stmt->element_type()), stmt->arg_id);
+    }
+  }
+
+  void visit(ArgStoreStmt *stmt) override
+  {
+    TI_ASSERT(!stmt->is_ptr);
+    emit("_arg_{}({}) = {};", data_type_short_name(stmt->element_type()),
+        stmt->arg_id, stmt->val->raw_name());
+  }
+
+  std::string make_kernel_name()
+  {
+    return fmt::format("{}{}", glsl_kernel_prefix_, glsl_kernel_count_++);
+  }
+
+  void generate_serial_kernel(OffloadedStmt *stmt) {
+    TI_ASSERT(stmt->task_type == OffloadedStmt::TaskType::serial);
+    const std::string glsl_kernel_name = make_kernel_name();
+    this->glsl_kernel_name_ = glsl_kernel_name;
+    emit("void {}()", glsl_kernel_name);
+    emit("{{ // serial");
+    stmt->body->accept(this);
+    emit("}}\n");
+  }
+
+
+  void visit(OffloadedStmt *stmt) override
+  {
+    TI_ASSERT(is_top_level_); // TODO(archibate): remove for nested kernel (?)
+    is_top_level_ = false;
+    using Type = OffloadedStmt::TaskType;
+    if (stmt->task_type == Type::serial) {
+      generate_serial_kernel(stmt);
+    /*} else if (stmt->task_type == Type::range_for) {
+      generate_range_for_kernel(stmt);*/
+    } else {
+      // struct_for is automatically lowered to ranged_for for dense snodes
+      // (#378). So we only need to support serial and range_for tasks.
+      TI_ERROR("Unsupported offload type={} on OpenGL arch", stmt->task_name());
+    }
+    is_top_level_ = true;
+  }
+
+
+public:
+  const std::string &kernel_source_code() const
+  {
+    return kernel_src_code_;
+  }
+
+  SSBO *create_root_ssbo()
```

Comment:
Why static? Could you use `std::unique_ptr<SSBO>` here?

Reply:
This is bad, I know. But don't worry, I will treat this in a more elegant way in the later stages. See 72a6005.